### DISCUSSION TITLE
Creating string of hashes from PVOID variable

### DISCUSSION THREAD

**Discussion by Airick73** [winsiderss/systeminformer#1489] [Questions / Answers] [answered]
I was trying to find a way to create a string of hashes from a PVOID variable in a for loop. For reference I have the following code
![image](https://user-images.githubusercontent.com/72322466/202053931-d7013af1-8562-432d-9499-1defeaafb7b6.png)

`resultTempList->Items[i]` is a PVOID*
`jsonString` is a PHH_STRING. 

Inside the `MDCBuildJsonArray` function I want to retrieve the hash from `resultTempList->Items[i]` and concatenate it to `jsonString`. Note I am able to successfully retrieve the hash with the following method. 
![image](https://user-images.githubusercontent.com/72322466/202054498-3c2c326a-86c4-4e5d-bb3e-e1e4be0da390.png)
My problem is finding a good function within Process Hacker that successfully concatenate a string. I've tried the functions such as `PPH_STRING PhConcatStrings2(
    _In_ PWSTR String1,
    _In_ PWSTR String2
    )` and `PPH_STRING PhConcatStringRef2(
    _In_ PPH_STRINGREF String1,
    _In_ PPH_STRINGREF String2
    )` but they seem to lead to errors which lead to no where for me such as [Ntdll.dll](https://www.lifewire.com/how-to-fix-ntdll-dll-errors-2624474). If anyone has experience with concatenating strings from different objects within Process Hacker and can share insight I would greatly appreciate it.

**Comment by Advait251206** [comment, marked as answer]
The string concatenation functions themselves may not be the problem here. Looking at the code snippet, I'd first verify that the types being passed around are what the Process Hacker string APIs expect.

A few observations:

### 1. `PhConcatStrings2` returns a new string

Functions such as:

```c
PPH_STRING PhConcatStrings2(
    _In_ PWSTR String1,
    _In_ PWSTR String2
);
```

don't append to an existing `PPH_STRING`; they allocate and return a new one.

For example:

```c
PPH_STRING newString = PhConcatStrings2(
    jsonString->Buffer,
    hashString->Buffer
);
```

You would then need to manage the references correctly:

```c
PhDereferenceObject(jsonString);
jsonString = newString;
```

If you ignore the returned value, nothing is actually appended.

### 2. Watch out for ownership/reference counting

Most `PPH_STRING` objects are reference-counted. If you're repeatedly concatenating strings inside a loop, you'll want to dereference temporary strings to avoid leaks.

Something like:

```c
PPH_STRING temp;

temp = PhConcatStrings2(
    jsonString->Buffer,
    hashString->Buffer
);

PhDereferenceObject(jsonString);
jsonString = temp;
```

### 3. Consider using a string builder

If you're building a large JSON string in a loop, repeatedly concatenating strings can become inefficient because every concatenation allocates a new buffer.

Process Hacker has string builder APIs that are generally better suited for this pattern. Depending on the version you're working with, look for structures/functions related to:

```c
PH_STRING_BUILDER
```

and operations such as:

```c
PhInitializeStringBuilder(...)
PhAppendStringBuilder(...)
PhAppendFormatStringBuilder(...)
```

Building the JSON incrementally is usually much cleaner than repeated concatenation.

### 4. Verify the cast from `PVOID`

You mention:

```c
resultTempList->Items[i]
```

is a `PVOID *`, while your function expects:

```c
PMDC_FILE_HASH_ENTRY Entry
```

Make sure the stored object really is a `PMDC_FILE_HASH_ENTRY` before passing it in:

```c
MDCBuildJsonArray(
    (PMDC_FILE_HASH_ENTRY)resultTempList->Items[i],
    jsonString
);
```

A bad cast or invalid pointer could easily lead to crashes that appear to originate in `ntdll.dll`.

### 5. JSON generation

Since you're ultimately generating JSON, you may find it easier to append directly to a builder:

```c
PhAppendStringBuilder2(&builder, L"\"");
PhAppendStringBuilder(&builder, hashString);
PhAppendStringBuilder2(&builder, L"\",");
```

rather than repeatedly creating new `PPH_STRING` instances.

Without the exact crash location it's hard to be certain, but if the failure occurs inside `ntdll.dll`, I'd be suspicious of either:

- An invalid `PPH_STRING` reference,
- A bad cast from `PVOID`,
- A string object being dereferenced after release,
- Or passing a null/invalid `Buffer` into the concatenation functions.

If you can share the exact call that crashes and the debugger stack trace, it would be much easier to pinpoint.
